django/db/models/fields/files.py

```diff
@@ -225,6 +225,10 @@ def __init__(self, verbose_name=None, name=None, upload_to='', storage=None, **k
         self._primary_key_set_explicitly = 'primary_key' in kwargs
 
         self.storage = storage or default_storage
+
+        if callable(self.storage):
+            self.storage = self.storage()
```

Comment:
I could add an issubclass, if needed, but the added overhead may not be necessary.